core/src/main/java/com/graphhopper/routing/ch/CHProfileSelector.java

```diff
@@ -0,0 +1,185 @@
+/*
+ *  Licensed to GraphHopper GmbH under one or more contributor
+ *  license agreements. See the NOTICE file distributed with this work for
+ *  additional information regarding copyright ownership.
+ *
+ *  GraphHopper GmbH licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except in
+ *  compliance with the License. You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.graphhopper.routing.ch;
+
+import com.carrotsearch.hppc.IntObjectHashMap;
+import com.carrotsearch.hppc.IntObjectMap;
+import com.graphhopper.routing.util.HintsMap;
+import com.graphhopper.storage.CHProfile;
+import com.graphhopper.util.Parameters;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * This class is used to determine the appropriate CH profile given (or not given) some (request) parameters
+ */
+public class CHProfileSelector {
+    private final List<CHProfile> chProfiles;
+    private HintsMap weightingMap;
+    private IntObjectMap<CHProfile> edgeBasedCHProfilesByUTurnCosts;
+    private CHProfile nodeBasedCHProfile;
+    private List<String> entriesStrs;
+
+    private CHProfileSelector(List<CHProfile> chProfiles) {
+        this.chProfiles = chProfiles;
+    }
+
+    /**
+     * @param chProfiles   the CH profiles to choose from
+     * @param weightingMap a map used to specify the weighting that shall be used
+     * @param edgeBased    whether or not edge-based CH shall be used or null if not specified explicitly
+     * @param uTurnCosts   specifies which value the u-turn costs of the CH profile should have, or null
+     */
+    public static CHProfile select(
+            List<CHProfile> chProfiles, HintsMap weightingMap, Boolean edgeBased, Integer uTurnCosts)
+            throws CHProfileSelectionException {
+        return new CHProfileSelector(chProfiles).select(weightingMap, edgeBased, uTurnCosts);
+    }
+
+    private CHProfile select(HintsMap weightingMap, Boolean edgeBased, Integer uTurnCosts) throws CHProfileSelectionException {
+        this.weightingMap = weightingMap;
+        findCHProfilesMatchingWeighting();
+
+        if (!foundCHProfilesMatchingWeighting()) {
+            throw new CHProfileSelectionException("Cannot find CH profile for weighting map " + weightingMap + " in entries: " + entriesStrs + ".");
+        }
+
+        if (edgeBased != null && uTurnCosts != null) {
```

Comment:
This is again quite a bit of nesting (even though its spread over different methods), but I thought this way it is easier to be sure we cover all the possible cases ? wdyt ?

Reply:
Looks fine due to the named cases IMO